gdnative-core/src/geom/basis.rs

```diff
@@ -29,14 +116,316 @@ impl Basis {
+        det.approx_eq(&1.0) && self.is_orthogonal()
+    }
+
+    /// Multiplies the matrix from left by the rotation matrix: M -> R.M
+    ///
+    /// The main use of Basis is as Transform.basis, which is used a the transformation matrix
+    /// of 3D object. Rotate here refers to rotation of the object (which is R * (*this)),
+    pub fn rotated(self, axis: &Vector3, phi: f32) -> Basis {
+        let rot = Basis::from_axis_angle(&axis, phi);
+        rot * self
+    }
+
+    /// Rotates the matrix.
+    ///
+    /// See [`Basis::rotated()`](./struct.Basis.html#method.rotated)
+    pub fn rotate(&mut self, axis: &Vector3, phi: f32) {
+        *self = self.rotated(axis, phi);
+    }
+
+    /// The basis’s rotation in the form of a quaternion.
+    ///
+    /// See [`Basis::euler()`](./struct.Basis.html#method.euler) if you need Euler angles, but keep in mind quaternions should generally be preferred to Euler angles.
+    pub fn into_quat(self) -> Quat {
+        assert!(self.is_rotation(), "Basis must be normalized in order to be casted to a Quaternion. Use get_rotation_quat() or call orthonormalized() instead.");
+
+        // Allow getting a quaternion from an unnormalized transform
+        let trace = self.elements[0].x + self.elements[1].y + self.elements[2].z;
+        let mut temp = [0_f32; 4];
+
+        if trace > 0.0 {
+            let mut s = (trace + 1.0).sqrt();
+            temp[3] = s * 0.5;
+            s = 0.5 / s;
+
+            temp[0] = (self.elements[2].y - self.elements[1].z) * s;
+            temp[1] = (self.elements[0].z - self.elements[2].x) * s;
+            temp[2] = (self.elements[1].x - self.elements[0].y) * s;
+        } else {
+            let i = if self.elements[0].x < self.elements[1].y {
+                if self.elements[1].y < self.elements[2].z {
+                    2
+                } else {
+                    1
+                }
+            } else {
+                if self.elements[0].x < self.elements[2].z {
+                    2
+                } else {
+                    0
+                }
+            };
+
+            let j = (i + 1) % 3;
+            let k = (i + 2) % 3;
+
+            let elements_arr: [[f32; 3]; 3] = [
+                self.elements[0].to_array(),
+                self.elements[1].to_array(),
+                self.elements[2].to_array(),
+            ];
+
+            let mut s = (elements_arr[i][i] - elements_arr[j][j] - elements_arr[k][k] + 1.0).sqrt();
+            temp[i] = s * 0.5;
+            s = 0.5 / s;
+
+            temp[3] = (elements_arr[k][j] - elements_arr[j][k]) * s;
+            temp[j] = (elements_arr[j][i] + elements_arr[i][j]) * s;
+            temp[k] = (elements_arr[k][i] + elements_arr[i][k]) * s;
         }
+
+        let [a, b, c, r] = temp;
+        Quat::quaternion(a, b, c, r)
+    }
+
+    /// Converts matrix into a Quaternion
+    ///
+    /// See [`Basis::into_quat`](./struct.Basis.html#method.into_quat)
+    pub fn quat(&self) -> Quat {
```

Comment:
Should be called `to_quat`, and contain the code currently in `into_quat`. The current `into_quat` should be removed.